### DISCUSSION TITLE
Optional path parameters in dependencies and OpenAPI

### DISCUSSION THREAD

**Discussion by mrlubos** [fastapi/fastapi#5459] [Questions] [answered]
Hey! I have a dependency that can handle path parameters if they exist. For example, `{id}` would trigger a side effect when it's defined (e.g. `GET /foo/{id}`), but the side effect wouldn't run if `{id}` isn't defined (e.g. `GET /foo`). To achieve this, I define `{id}` in my dependency as follows: `id: Optional[int] = None`. This works great.

However, the OpenAPI specs now contain the `id` parameter for every path operation using my dependency. I don't want this, since every path operation contains its own parameters. The dependency should just handle them if they're defined. Is this possible to achieve?

**Comment by mrlubos** [comment]
I tried doing `id: Optional[int] = Path(include_in_schema=False)` but that excludes it from every path operation. Ideally, there would be a way to set this argument based on the endpoint path.

**Comment by vanishingESCkey** [comment]
Not sure if this fits your needs but you can wrap dependency in a method used to decide if id should be used or not. The OpenAPI shows id is required only for `use_id=True` However I would rather just use two different dependencies in this case:

 ```python
from typing import Callable, Optional

from fastapi import Depends, FastAPI

app = FastAPI()


def generate_message(id: Optional[int] = None) -> str:
    return "no id" if id is None else f"got id {id}"

def build_dep(use_id: Optional[bool] = False) -> Callable:
    def dep_without_id() -> str:
        return generate_message()

    def dep_with_id(id: int) -> str:
        return generate_message(id)

    return dep_with_id if use_id else dep_without_id


@app.get("/foo")
async def foo(message: str = Depends(build_dep(use_id=True))):
    return {"message": message}


@app.get("/bar")
async def bar(message: str = Depends(build_dep(use_id=False))):
    return {"message": message}
```

**Reply by mrlubos** [reply]
Hi @vanishingESCkey, thank you for the reply! Reading your example, it sounds like it should work, thanks!

However, I still feel like this should be handled at the library level. It's extra boilerplate to work around something I consider a bug. Is it really not possible to achieve what I'm trying to do without this boilerplate?

The issue seems simple. When you have an optional path parameter `foo`, don't include it in the OpenAPI schema if there's no `foo` parameter in the path route.

The problem is my current implementation will in that case fallback to query parameters and `foo` gets included as an optional query parameter. This is not what I want, so I explicitly define it as `foo = Path()`. This, however, doesn't respect the `Optional[int]` type hint and turns `foo` into required everywhere.

It is for that reason that I think this is a bug. If I define my parameter as `foo: Optional[int] = Path()`, it should not be included (required) if the path doesn't contain it. Would you agree?

**Comment by nymous** [comment, marked as answer]
In OpenAPI, path parameters are always required (https://swagger.io/docs/specification/describing-parameters/#path-parameters). The specification doesn't allow you to work around that.
As another solution, you could use the `request` object directly (see https://fastapi.tiangolo.com/advanced/using-request-directly/ and https://www.starlette.io/requests/#path-parameters):
```py
from fastapi import FastAPI, Depends, Path, Request

app = FastAPI()


def get_optional_id(request: Request) -> str:
    id_ = request.path_params.get("id")
    if id_ is None:
        return "id was NOT provided"
    return "id was provided"

@app.get("/withoutId")
def get_without_id(message: str = Depends(get_optional_id)):
    print(message)

@app.get("/withId/{id}")
def get_with_id(id_: str = Path(..., alias="id"), message: str = Depends(get_optional_id)):
    print(message)
```

**Reply by mrlubos** [reply]
Thank you @nymous! I'm clearly not very familiar with the OpenAPI specs. Now that you've explained that requirement, it all makes sense. I can confirm using the request object directly instead of defining my own parameters produces the desired results. Thanks again!
